File: backend/app/schemas/boost.py
"""
app/schemas/boost.py

Request/response models for the boost-plan catalog and boost-order
checkout — backs the broker "Boost Listing" page.
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.enums import BoostTier, OrderStatus

# The only durations the checkout page offers; each carries its own
# volume discount (see boost_service.DURATION_DISCOUNT_PERCENT).
BoostDurationDays = Literal[7, 15, 30]


class ReachEstimate(BaseModel):
    """Illustrative views/leads/calls ranges for a plan's "Expected Reach" panel."""

    views_min: int
    views_max: int
    leads_min: int
    leads_max: int
    calls_min: int
    calls_max: int


class BoostPlanRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    tier: BoostTier
    price: float
    features: list[str]
    reach_estimate: ReachEstimate


class BoostOrderCreateRequest(BaseModel):
    property_id: UUID
    plan_id: UUID
    duration_days: BoostDurationDays


class BoostOrderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    property_id: UUID
    plan_id: UUID
    duration_days: int
    base_amount: float
    discount_amount: float
    gst_amount: float
    total_amount: float
    status: OrderStatus
    created_at: datetime

File: frontend/lib/api/endpoints/boost.ts
// lib/api/endpoints/boost.ts
// Boost-plan catalog (public) and boost-order checkout (broker-only) —
// backs the Boost Listing page. No payment gateway is wired yet; an
// order is created and stays in "created" status.

import { apiRequest } from "@/lib/api/client";
import type { BoostTier, OrderStatus } from "@/lib/enums";

export type ReachEstimate = {
  views_min: number;
  views_max: number;
  leads_min: number;
  leads_max: number;
  calls_min: number;
  calls_max: number;
};

export type BoostPlan = {
  id: string;
  name: string;
  tier: BoostTier;
  price: number;
  features: string[];
  reach_estimate: ReachEstimate;
};

export function listBoostPlans(): Promise<BoostPlan[]> {
  return apiRequest<BoostPlan[]>("/boost-plans");
}

export type BoostOrderCreateInput = {
  property_id: string;
  plan_id: string;
  duration_days: 7 | 15 | 30;
};

export type BoostOrder = {
  id: string;
  property_id: string;
  plan_id: string;
  duration_days: number;
  base_amount: number;
  discount_amount: number;
  gst_amount: number;
  total_amount: number;
  status: OrderStatus;
  created_at: string;
};

export function createBoostOrder(data: BoostOrderCreateInput): Promise<BoostOrder> {
  return apiRequest<BoostOrder>("/boost-orders", { method: "POST", body: data });
}
